devito/types/basic.py

```diff
@@ -382,7 +386,8 @@ def __eq__(self, other):
     __hash__ = sympy.Symbol.__hash__
 
     def _hashable_content(self):
-        return super()._hashable_content() + (self.dtype, self.is_const)
+        return super()._hashable_content() \
+            + (self.dtype, self.is_const, self.nonnegative)
```

Comment:
isn't it part of super() already? can you add a test to confirm this. 

All the assumptions should be  part of the hash clearly, but we don't want to have to list them here explicitly one by one